geopandas/io/sql.py

```diff
@@ -1,46 +1,65 @@
-import binascii
-
 from pandas import read_sql
 import shapely.wkb
 
 from geopandas import GeoSeries, GeoDataFrame
 
 
-def read_postgis(sql, con, geom_col='geom', crs=None, index_col=None,
-                 coerce_float=True, params=None):
+def read_postgis(sql, con, geom_col='geom', crs=None, hex_encoded=True,
+                 index_col=None, coerce_float=True, params=None):
     """
     Returns a GeoDataFrame corresponding to the result of the query
     string, which must contain a geometry column.
 
-    Examples:
-    sql = "SELECT geom, kind FROM polygons;"
-    df = geopandas.read_postgis(sql, con)
-
     Parameters
     ----------
-    sql: string
-    con: DB connection object or SQLAlchemy engine
-    geom_col: string, default 'geom'
+    sql : string
+        SQL query to execute in selecting entries from database.
+    con : DB connection object or SQLAlchemy engine
+        Active connection to the database to query.
+    geom_col : string, default 'geom'
         column name to convert to shapely geometries
-    crs: optional
-        CRS to use for the returned GeoDataFrame
+    crs : dict or str, optional
+        CRS to use for the returned GeoDataFrame; if not set, tries to
+        determine CRS from the SRID associated with geometries in the
+        database.
+    hex_encoded : bool, optional
+        Whether the geometry is in a hex-encoded string. Default is True,
+        standard for postGIS.
 
     See the documentation for pandas.read_sql for further explanation
     of the following parameters:
     index_col, coerce_float, params
 
+    Returns
+    -------
+    geodataframe : GeoDataFrame
+
+    Example
+    -------
+    >>> sql = "SELECT geom, kind FROM polygons;"
+    >>> df = geopandas.read_postgis(sql, con)
     """
+
     df = read_sql(sql, con, index_col=index_col, coerce_float=coerce_float,
                   params=params)
 
     if geom_col not in df:
-        raise ValueError("Query missing geometry column '{0}'".format(
-            geom_col))
+        raise ValueError("Query missing geometry column '{}'".format(geom_col))
 
+    def load_geom(x):
+        return shapely.wkb.loads(str(x), hex=hex_encoded)
     wkb_geoms = df[geom_col]
+    df[geom_col] = GeoSeries(wkb_geoms.apply(load_geom))
 
-    s = wkb_geoms.apply(lambda x: shapely.wkb.loads(binascii.unhexlify(x.encode())))
-
-    df[geom_col] = GeoSeries(s)
+    if crs is None:
+        def get_srid(x):
+            return shapely.geos.lgeos.GEOSGetSRID(x._geom)
+        unique_srids = df[geom_col].apply(get_srid).unique()
```

Comment:
I most cases there will only be one SRID, so it could be sufficient to only get the srid from the first.. (which would be more performant)

It's only a bit unfortunate that postgis allows multiple srids if the srid is not set for the column.  
But on the other hand, it is really bad practice and your resulting GeoDataFrame is practically unusable if you are mixing srids? 
Or are there good use cases for this?

(we could also document that we assume a unique SRID. In the end, it is not changing any data, it is just the `.crs` attribute that is then not fully correct)

Reply:
My thought on this was that leaving an empty CRS in this case was the least bad option, but didn't give consideration to any performance hit. I think taking the SRID from the first geometry would be a totally reasonable option though, so long as we warn users, so if you think that's the better way to go, happy to modify this accordingly.

Reply:
I am not sure, but it seems sensible to me yes (if you have a mix, the .crs attribute is not useful anyway)

Reply:
Right, that's why in the implementation I have now I chose to leave the CRS as `None`, rather than set the CRS to the value from the first geometry, as that seems less likely to produce any surprising behavior. I don't even know how frequently this situation would arise though.

Reply:
Given that it is probably a rare situation, I would go for checking only the first and putting a note about that in the docs (will also make the code a bit simpler)